### ISSUE TITLE
Can’t mount by NFS vagrant 2.3.4

### ISSUE THREAD

**Issue by markrusselldev** [hashicorp/vagrant#13078]
I am getting an error when trying vagrant up. I have tried at least 10 different "fixes" from online articles.  Nothing works.

### Debug output
```
The following SSH command responded with a non-zero exit status.
Vagrant assumes that this means the command failed!

mount -o vers=3,udp 192.168.56.1:/home/mark/Dev/Trellis/learn-trellis/site /vagrant-nfs-example.com

Stdout from the command:



Stderr from the command:

mount.nfs: Connection timed out
```

### Expected behavior

It should complete vagrant up without issue.

### Actual behavior

mount.nfs: Connection timed out

### Reproduction information 

run vagrant up

#### Vagrant version

Vagrant 2.3.4

#### Host operating system

Kubuntu 22.04

#### Guest operating system

bento/ubuntu-22.04

#### Steps to reproduce

1. Install vagrant
2. Install virtualbox
3. run vagrant up

#### Vagrantfile

```ruby
# Copy-paste your Vagrantfile here. Remove any sensitive information such as passwords, authentication tokens, or email addresses.
```
I have no idea what would be relevant information. Here's the Vagrantfile:
`ANSIBLE_PATH = __dir__ # absolute path to Ansible directory on host machine
ANSIBLE_PATH_ON_VM = '/home/vagrant/trellis'.freeze # absolute path to Ansible directory on virtual machine

require File.join(ANSIBLE_PATH, 'lib', 'trellis', 'vagrant')
require File.join(ANSIBLE_PATH, 'lib', 'trellis', 'config')
require 'yaml'

vconfig = YAML.load_file("#{ANSIBLE_PATH}/vagrant.default.yml")

if File.exist?("#{ANSIBLE_PATH}/vagrant.local.yml")
  local_config = YAML.load_file("#{ANSIBLE_PATH}/vagrant.local.yml")
  vconfig.merge!(local_config) if local_config
end

ensure_plugins(vconfig.fetch('vagrant_plugins')) if vconfig.fetch('vagrant_install_plugins')

trellis_config = Trellis::Config.new(root_path: ANSIBLE_PATH)

Vagrant.require_version vconfig.fetch('vagrant_require_version', '>= 2.1.0')

Vagrant.configure('2') do |config|
  box = vconfig.fetch('vagrant_box')
  box_auto_arch = vconfig.fetch('vagrant_box_auto_arch', true)

  if box_auto_arch && !box.end_with?("-arm64") && apple_silicon?
    box = "#{box}-arm64"
  end

  config.vm.box = box
  config.vm.box_version = vconfig.fetch('vagrant_box_version')
  config.ssh.forward_agent = true
  config.vm.post_up_message = post_up_message

  # Fix for: "stdin: is not a tty"
  # https://github.com/mitchellh/vagrant/issues/1673#issuecomment-28288042
  config.ssh.shell = %(bash -c 'BASH_ENV=/etc/profile exec bash')

  # Required for NFS to work
  if vconfig.fetch('vagrant_ip') == 'dhcp'
    config.vm.network :private_network, type: 'dhcp', hostsupdater: 'skip'

    cached_addresses = {}
    config.hostmanager.ip_resolver = proc do |vm|
      if cached_addresses[vm.name].nil?
        if vm.communicate.ready?
          vm.communicate.execute("hostname -I | cut -d ' ' -f 2") do |_type, contents|
            cached_addresses[vm.name] = contents.split("\n").first[/(\d+\.\d+\.\d+\.\d+)/, 1]
          end
        end
      end
      cached_addresses[vm.name]
    end
  else
    config.vm.network :private_network, ip: vconfig.fetch('vagrant_ip'), hostsupdater: 'skip'
  end

  main_hostname, *hostnames = trellis_config.site_hosts_canonical
  config.vm.hostname = main_hostname

  if Vagrant.has_plugin?('vagrant-hostmanager') && !trellis_config.multisite_subdomains?
    redirects = trellis_config.site_hosts_redirects

    config.hostmanager.enabled = true
    config.hostmanager.manage_host = true
    config.hostmanager.aliases = hostnames + redirects
  elsif Vagrant.has_plugin?('landrush') && trellis_config.multisite_subdomains?
    config.landrush.enabled = true
    config.landrush.tld = trellis_config.site_hosts_canonical.reject { |host| host.end_with?(".#{main_hostname}") }
    hostnames.each { |host| config.landrush.host host, vconfig.fetch('vagrant_ip') }
  else
    fail_with_message "vagrant-hostmanager missing, please install the plugin with this command:\nvagrant plugin install vagrant-hostmanager\n\nOr install landrush for multisite subdomains:\nvagrant plugin install landrush"
  end

  vagrant_mount_type = vconfig.fetch('vagrant_mount_type')

  extra_options = if vagrant_mount_type == 'smb'
    {
      smb_username: vconfig.fetch('vagrant_smb_username', nil),
      smb_password: vconfig.fetch('vagrant_smb_password', nil),
    }
  else
    {}
  end

  if vagrant_mount_type != 'nfs' || Vagrant::Util::Platform.wsl? || (Vagrant::Util::Platform.windows? && !Vagrant.has_plugin?('vagrant-winnfsd'))
    vagrant_mount_type = nil if vagrant_mount_type == 'nfs'
    trellis_config.wordpress_sites.each_pair do |name, site|
      config.vm.synced_folder local_site_path(site), remote_site_path(name, site), owner: 'vagrant', group: 'www-data', mount_options: mount_options(vagrant_mount_type, dmode: 776, fmode: 775), type: vagrant_mount_type, **extra_options
    end

    config.vm.synced_folder ANSIBLE_PATH, ANSIBLE_PATH_ON_VM, mount_options: mount_options(vagrant_mount_type, dmode: 755, fmode: 644), type: vagrant_mount_type, **extra_options
  elsif !Vagrant.has_plugin?('vagrant-bindfs')
    fail_with_message "vagrant-bindfs missing, please install the plugin with this command:\nvagrant plugin install vagrant-bindfs"
  else
    trellis_config.wordpress_sites.each_pair do |name, site|
      config.vm.synced_folder local_site_path(site), nfs_path(name), type: 'nfs'
      config.bindfs.bind_folder nfs_path(name), remote_site_path(name, site), u: 'vagrant', g: 'www-data', o: 'nonempty'
    end

    config.vm.synced_folder ANSIBLE_PATH, '/ansible-nfs', type: 'nfs'
    config.bindfs.bind_folder '/ansible-nfs', ANSIBLE_PATH_ON_VM, o: 'nonempty', p: '0644,a+D'
  end

  vconfig.fetch('vagrant_synced_folders', []).each do |folder|
    options = {
      type: folder.fetch('type', 'nfs'),
      create: folder.fetch('create', false),
      mount_options: folder.fetch('mount_options', [])
    }

    destination_folder = folder.fetch('bindfs', true) ? nfs_path(folder['destination']) : folder['destination']

    config.vm.synced_folder folder['local_path'], destination_folder, options

    if folder.fetch('bindfs', true)
      config.bindfs.bind_folder destination_folder, folder['destination'], folder.fetch('bindfs_options', {})
    end
  end

  provisioner = local_provisioning? ? :ansible_local : :ansible
  provisioning_path = local_provisioning? ? ANSIBLE_PATH_ON_VM : ANSIBLE_PATH

  config.vm.provision provisioner do |ansible|
    if local_provisioning?
      ansible.install_mode = 'pip'
      ansible.pip_install_cmd = 'sudo apt-get install -y -qq python3-pip'
      ansible.provisioning_path = provisioning_path
      ansible.version = vconfig.fetch('vagrant_ansible_version')
    end

    ansible.compatibility_mode = '2.0'
    ansible.playbook = File.join(provisioning_path, 'dev.yml')
    ansible.galaxy_role_file = File.join(provisioning_path, 'galaxy.yml') unless vconfig.fetch('vagrant_skip_galaxy') || ENV['SKIP_GALAXY']
    ansible.galaxy_roles_path = File.join(provisioning_path, 'vendor/roles')

    if which('trellis')
      ansible.galaxy_command = 'trellis galaxy install'
    end

    ansible.groups = {
      'web' => ['default'],
      'development' => ['default']
    }

    ansible.tags = ENV['ANSIBLE_TAGS']
    ansible.extra_vars = { 'vagrant_version' => Vagrant::VERSION }

    if (vars = ENV['ANSIBLE_VARS'])
      extra_vars = Hash[vars.split(',').map { |pair| pair.split('=') }]
      ansible.extra_vars.merge!(extra_vars)
    end

    if !Vagrant::Util::Platform.windows?
      config.trigger.after :up do |trigger|
        # Add Vagrant ssh-config to ~/.ssh/config
        trigger.info = "Adding vagrant ssh-config for #{main_hostname } to ~/.ssh/config"
        trigger.ruby do
          update_ssh_config(main_hostname)
        end
      end
    end
  end

  # VirtualBox settings
  config.vm.provider 'virtualbox' do |vb|
    vb.name = config.vm.hostname
    vb.customize ['modifyvm', :id, '--cpus', vconfig.fetch('vagrant_cpus')]
    vb.customize ['modifyvm', :id, '--memory', vconfig.fetch('vagrant_memory')]
    vb.customize ['modifyvm', :id, '--ioapic', vconfig.fetch('vagrant_ioapic', 'on')]

    # Fix for slow external network connections
    vb.customize ['modifyvm', :id, '--natdnshostresolver1', vconfig.fetch('vagrant_natdnshostresolver', 'on')]
    vb.customize ['modifyvm', :id, '--natdnsproxy1', vconfig.fetch('vagrant_natdnsproxy', 'on')]
  end

  # VMware Workstation/Fusion settings
  %w(vmware_fusion vmware_workstation).each do |provider|
    config.vm.provider provider do |vmw|
      vmw.vmx['displayName'] = config.vm.hostname
      vmw.vmx['numvcpus'] = vconfig.fetch('vagrant_cpus')
      vmw.vmx['memsize'] = vconfig.fetch('vagrant_memory')
    end
  end

  # Parallels settings
  config.vm.provider 'parallels' do |prl|
    prl.name = config.vm.hostname
    prl.cpus = vconfig.fetch('vagrant_cpus')
    prl.memory = vconfig.fetch('vagrant_memory')
    prl.update_guest_tools = true

    # Parallels handles DNS resolution itself when used in conjunction with landrush
    if Vagrant.has_plugin?('landrush') && trellis_config.multisite_subdomains?
      config.landrush.guest_redirect_dns = false
    end
  end

  # Hyper-V settings
  config.vm.provider 'hyperv' do |h|
    h.vmname = config.vm.hostname
    h.cpus = vconfig.fetch('vagrant_cpus')
    h.memory = vconfig.fetch('vagrant_memory')
    h.enable_virtualization_extensions = true
    h.linked_clone = true
  end
end

 # https://developer.hashicorp.com/vagrant/docs/synced-folders/nfs#prerequisites
Vagrant.configure("2") do |config|
  config.vm.network "private_network", type: "dhcp"
end

Vagrant.configure("2") do |config|
  config.vm.synced_folder ".", "/vagrant", type: "nfs"
end
`

And here's vagrant.default.yml

`---
vagrant_ip: '192.168.56.5'
vagrant_cpus: 1
vagrant_memory: 1024 # in MB
vagrant_box: 'bento/ubuntu-22.04'
vagrant_box_version: '>= 0'
vagrant_box_auto_arch: true
vagrant_ansible_version: '2.10.7'
vagrant_skip_galaxy: false
vagrant_mount_type: 'nfs'
vagrant_require_version: '>= 2.1.0'

vagrant_install_plugins: true
vagrant_plugins:
  - name: vagrant-bindfs
  - name: vagrant-hostmanager

# Array of synced folders:
# - local_path: .
#   destination: /path/on/vm
#   create: false
#   type: nfs
#   bindfs: true
#   mount_options: []
#   bindfs_options: {}
# See https://www.vagrantup.com/docs/synced-folders/basic_usage.html#mount_options
vagrant_synced_folders: []

vagrant ALL=(ALL) NOPASSWD: ALL
`

**Comment by soapy1** [comment]
Hey there, thanks for opening up an issue. NFS issues are pretty challenging to resolve since they are pretty dependent on the host system setup. Since this is a connection error `mount.nfs: Connection timed out` it looks like some networking is not quite in place. To get a better understanding of what is going wrong could you share

1. the minimal vagrantfile for reproducing this issue (the current provided Vagrantfile is hard to parse and looks like it might have some irrelevant details)
2. the full [debug output from vagrant](https://developer.hashicorp.com/vagrant/docs/other/debugging)
3. if possible, the most verbose logs from manually trying to mount the file (executing the command that Vagrant said failed)

Thanks!

**Comment by zekefarwell** [comment]
I'm running into similar issues.  Most likely a duplicate of #12758.  It seems that in version 22.04, Ubuntu has [disabled using UDP as the transport for NFS mounts](https://discourse.ubuntu.com/t/jammy-jellyfish-release-notes/24668), regardless of NFS version.  Since Vagrant defaults to UDP for NFS mounts, this causes an error.  

A workaround is to [set nfs_udp=false on the synced folder](https://developer.hashicorp.com/vagrant/docs/synced-folders/nfs#nfs-synced-folder-options) which causes TCP to be used instead.  Unfortunately with the application I'm running, I'm seeing much worse performance using TCP than UDP, so this isn't a great solution for me.  I'm hoping this [Ubuntu issue report](https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1964093) will be considered as a bug rather than working as intended.

**Comment by poppygrunfeld** [comment]
> I'm running into similar issues. Most likely a duplicate of #12758. It seems that in version 22.04, Ubuntu has [disabled using UDP as the transport for NFS mounts](https://discourse.ubuntu.com/t/jammy-jellyfish-release-notes/24668), regardless of NFS version. Since Vagrant defaults to UDP for NFS mounts, this causes an error.
> 
> A workaround is to [set nfs_udp=false on the synced folder](https://developer.hashicorp.com/vagrant/docs/synced-folders/nfs#nfs-synced-folder-options) which causes TCP to be used instead. Unfortunately with the application I'm running, I'm seeing much worse performance using TCP than UDP, so this isn't a great solution for me. I'm hoping this [Ubuntu issue report](https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1964093) will be considered as a bug rather than working as intended.

This situation seems to come from defaults for NFS that I found confusing.

https://developer.hashicorp.com/vagrant/docs/synced-folders/nfs#other-notes

I am noticing that the default share (`/vagrant`) is NFSv4.2 when I check with `mount`, but my extra added share (that I used the above tip for) was coming up as NFSv3.

This is working for me,

```
  config.vm.synced_folder "../", "/srv/repo",
    type: "nfs",
    nfs_udp: false,
    nfs_version: 4,
    mount_options: ["ro"]
```

But it might be reasonable if these were the defaults for `nfs_udp` and `nfs_version`, at least for modern linux.
